File: src/deque/linked-deque.cpp
// - - - - - 
// Implementation of a deque (double queue) using a doubly linked list
// - - - - -

#include <iostream>
#include <exception>

struct node {
    node* prev;
    node* next;
    int val;
};

class dequeu {
  private:
    int size;
    node* first;
    node* last;

    void init(int value) {
        size = 1;
        node* ptr = new node();
        first = ptr;
        last = ptr;
        ptr->val = value;
    }

  public:
    dequeu() {
        size = 0;
        first = NULL;
        last = NULL;
    }

    void insertLast(int value) {
        if (isEmpty()) {
            init(value);
            return;
        }
        node* ptr = new node();
        ptr->val = value;
        ptr->prev = last;
        last->next = ptr;
        last = ptr;
        size++;
    }

    int removeLast() {
        if (isEmpty()) {
            throw new std::exception();
        }
        node* ptr = last;
        last = last->prev;
        size--;
        return ptr->val;
    }

    void insertFirst(int value) {
        if (isEmpty()) {
            init(value);
            return;
        }
        node* ptr = new node();
        ptr->val = value;
        ptr->next = first;
        first->prev = ptr;
        first = ptr;
        size++;
    }

    int removeFirst() {
        if (isEmpty()) {
            throw new std::exception();
        }
        node* ptr = first;
        first = first->next;
        size--;
        return ptr->val;
    }

    int count() {
        return size;
    }

    bool isEmpty() {
        return size == 0;
    }

    void print() {
        node* ptr = first;
        for (int i = 0; i < size; i++) {
            std::cout<<ptr->val<<' ';
            ptr = ptr->next;
        }
        std::cout<<std::endl;
    }
};

int main() {

    std::cout<<"Hello"<<std::endl;

    dequeu* d = new dequeu();

    d->insertLast(1);
    d->insertLast(2);
    d->insertLast(3);
    d->insertFirst(4);
    d->print();
    d->removeLast();
    d->print();
    d->removeFirst();
    d->print();
}

File: src/deque/LinkedDeque.java

/**
 *
 */
public class LinkedDeque {

    private static class Node {
        public Node prev;
        public Node next;
        public Object val;
    }

    private static class Deque {

        private Node first;
        private Node last;
        private int size;

        private void init(Object obj) {
            size = 1;
            Node node = new Node();
            node.val = obj;
            first = node;
            last = node;
        }



        public void insertFirst(Object val) {
            if (size == 0) {
                init(val);
                return;
            }
            Node node = new Node();
            node.val = val;
            node.next = first;
            first.prev = node;
            first = node;
            size++;
        }


        public void insertLast(Object val) {
            if (size == 0) {
                init(val);
                return;
            }
            Node node = new Node();
            node.val = val;
            node.prev = last;
            last.next = node;
            last = node;
            size++;
        }

        @Override
        public String toString() {
            String list = "";
            Node node = first;
            for (int i = 0; i < size; i++) {
                list += node.val.toString() + " ";
                node = node.next;
            }
            return list;
        }
    }

    public static void main(String[] args) {
        System.out.println("Hello");
        Deque deque = new Deque();
        deque.insertFirst(1);
        deque.insertFirst(2);
        deque.insertFirst(3);
        deque.insertLast(4);
        System.out.println(deque.toString());
    }
}
